_extensions/custom_roles.py

```diff
@@ -19,12 +25,16 @@ def setup(app):
 
 def autolink(pattern):
     def role(name, rawtext, text, lineno, inliner, options={}, content=[]):
+        """Combine literal + reference (unless the text is prefixed by a !)."""
         if " " in text:
             url_text = urllib.parse.quote(f"{text}")
         else:
             url_text = text
         url = pattern % (url_text,)
```

Comment:
```suggestion
        url = pattern.format(url_text)
```

Maybe worth using modern format syntax rather than old-style printf?